learning_observer/learning_observer/lms_integration.py

```diff
@@ -0,0 +1,287 @@
+import json
+import recordclass
+import string
+import aiohttp
+import aiohttp.web
+
+import learning_observer
+import learning_observer.runtime
+import learning_observer.google
+import learning_observer.constants as constants
+import learning_observer.settings as settings
+import pmss
+
+pmss.register_field(
+    name="lms_api",
+    type=pmss.pmsstypes.TYPES.string,
+    description="The Canvas Base API URL",
+    required=True
+)
+
+cache = None
+
+class Endpoint(recordclass.make_dataclass("Endpoint", ["name", "remote_url", "doc", "cleaners", "lms"], defaults=["", None])):
+    def arguments(self):
+        return extract_parameters_from_format_string(self.remote_url)
+
+    def _local_url(self):
+        parameters = "}/{".join(self.arguments())
+        base_url = f"/{self.lms}/{self.name}"
+        if len(parameters) == 0:
+            return base_url
+        else:
+            return base_url + "/{" + parameters + "}"
+
+    def _add_cleaner(self, name, cleaner):
+        if self.cleaners is None:
+            self.cleaners = dict()
+        self.cleaners[name] = cleaner
+        if 'local_url' not in cleaner:
+            cleaner['local_url'] = self._local_url + "/" + name
+
+    def _cleaners(self):
+        if self.cleaners is None:
+            return []
+        else:
+            return self.cleaners
+
+def extract_parameters_from_format_string(format_string):
+    '''
+    Extracts parameters from a format string. E.g.
+    >>> ("hello {hi} my {bye}")]
+    ['hi', 'bye']
+    '''
+    return [f[1] for f in string.Formatter().parse(format_string) if f[1] is not None]
+
+def raw_access_partial(raw_ajax_function, target_url, name=None):
+    '''
+    This is a helper which allows us to create a function which calls specific
+    Google APIs.
+
+    To test this, try:
+
+        print(await raw_document(request, documentId="some_google_doc_id"))
+    '''
+    async def caller(request, **kwargs):
+        '''
+        Make an AJAX request to LMS
+        '''
+        return await raw_ajax_function(request, target_url, **kwargs)
+    setattr(caller, "__qualname__", name)
+
+    return caller
+
+def api_docs_handler(endpoints):
```

Comment:
What is the security model here? Shouldn't these check if the user is authenticated or similar?